### DISCUSSION TITLE
GPU support

### DISCUSSION THREAD

**Discussion by ekwan** [grimme-lab/xtb#396] [Q&A]
Would xtb be accelerated using CUDA?  That would be really nice for using xtb for dynamics.  Thanks for making a great program!

**Comment by awvwgk** [comment]
`xtb` can already be accelerated using OpenACC (which is generating CUDA).

**Reply by ekwan** [reply]
Amazing! Could you please point to some documentation about how to do this?  Thanks.

**Reply by awvwgk** [reply]
Documentation has yet to be written (note to myself), it boils down to compiling `xtb` with the NVHPC compilers:
```
FC=nvfortran CC=nvc meson setup _build_gpu -Dgpu=true -Dcusolver=true -Dgpu_arch=70
```
If NVHPC 20.9 still generates invalid LLVM IR in `src/hessian.F90` the OpenMP directives have to be removed first.

**Reply by Chemistza74** [reply]
I have beet trying to compile the latest version of XTB with GPU support using nvhpc 20.11, but get the following error: 

> ERROR: Linker pgi does not support link_whole

My compile options are as follows:

> export FC=nvfortran CC=nvc
> 
> meson setup build_gpu -Dla_backend=openblas -Dopenmp=false -Dgpu=true -Dcusolver=true -Dgpu_arch=70 --prefix=$HOME/codes/xtb_GPU
> ninja -C build install

Output from the compilation log:

<details>

> The Meson build system
> Version: 0.56.0
> Source dir: /home/chris/Downloads/xtb/xtb
> Build dir: /home/chris/Downloads/xtb/xtb/build_gpu
> Build type: native build
> Project name: xtb
> Project version: 6.3.3
> Using 'FC' from environment with value: 'nvfortran'
> Fortran compiler for the host machine: nvfortran (nvidia_hpc 20.11-0)
> Fortran linker for the host machine: nvfortran pgi 20.11-0
> Using 'CC' from environment with value: 'nvc'
> C compiler for the host machine: nvc (nvidia_hpc 20.11-0)
> C linker for the host machine: nvc pgi 20.11-0
> Using 'FC' from environment with value: 'nvfortran'
> Using 'CC' from environment with value: 'nvc'
> Host machine cpu family: x86_64
> Host machine cpu: x86_64
> Program git found: YES (/usr/bin/git)
> Program date found: YES (/usr/bin/date)
> Program whoami found: YES (/usr/bin/whoami)
> Program hostname found: YES (/usr/bin/hostname)
> Configuring xtb_version.fh using configuration
> Found pkg-config: /usr/bin/pkg-config (0.29.2)
> Run-time dependency openblas found: YES 0.3.13.dev
> Found CMake: /usr/bin/cmake (3.16.3)
> Run-time dependency lapack found: NO (tried pkgconfig and cmake)
> Library lapack found: YES
> Run-time dependency threads found: YES
> Program asciidoctor found: NO
> Configuring config_env.bash using configuration
> Configuring config_env.csh using configuration
> Configuring xtb.pc using configuration
> Configuring xtb.tcl using configuration
> TESTSUITE/meson.build:52: DEPRECATION: ":" is not allowed in test name "Argparser: print version", it has been replaced with "_"
> TESTSUITE/meson.build:53: DEPRECATION: ":" is not allowed in test name "Argparser: print help", it has been replaced with "_"
> TESTSUITE/meson.build:54: DEPRECATION: ":" is not allowed in test name "Argparser: print license", it has been replaced with "_"
> TESTSUITE/meson.build:55: DEPRECATION: ":" is not allowed in test name "Argparser: no arguments", it has been replaced with "_"
> TESTSUITE/meson.build:77: DEPRECATION: ":" is not allowed in test name "Molecule: axis", it has been replaced with "_"
> TESTSUITE/meson.build:78: DEPRECATION: ":" is not allowed in test name "Molecule: MIC", it has been replaced with "_"
> TESTSUITE/meson.build:79: DEPRECATION: ":" is not allowed in test name "Lattice points: PBC-3D", it has been replaced with "_"
> TESTSUITE/meson.build:106: DEPRECATION: ":" is not allowed in test name "PBC tools: convert", it has been replaced with "_"
> TESTSUITE/meson.build:107: DEPRECATION: ":" is not allowed in test name "PBC tools: cutoff", it has been replaced with "_"
> TESTSUITE/meson.build:109: DEPRECATION: ":" is not allowed in test name "Symmetry: Water", it has been replaced with "_"
> TESTSUITE/meson.build:110: DEPRECATION: ":" is not allowed in test name "Symmetry: Li8", it has been replaced with "_"
> TESTSUITE/meson.build:111: DEPRECATION: ":" is not allowed in test name "Symmetry: PCl3", it has been replaced with "_"
> TESTSUITE/meson.build:112: DEPRECATION: ":" is not allowed in test name "Symmetry: large", it has been replaced with "_"
> TESTSUITE/meson.build:114: DEPRECATION: ":" is not allowed in test name "Thermo: axis", it has been replaced with "_"
> TESTSUITE/meson.build:115: DEPRECATION: ":" is not allowed in test name "Thermo: calculation", it has been replaced with "_"
> TESTSUITE/meson.build:116: DEPRECATION: ":" is not allowed in test name "Thermo: print", it has been replaced with "_"
> TESTSUITE/meson.build:118: DEPRECATION: ":" is not allowed in test name "Repulsion: 0D", it has been replaced with "_"
> TESTSUITE/meson.build:119: DEPRECATION: ":" is not allowed in test name "Repulsion: 3D", it has been replaced with "_"
> TESTSUITE/meson.build:121: DEPRECATION: ":" is not allowed in test name "Coulomb: point (0D)", it has been replaced with "_"
> TESTSUITE/meson.build:122: DEPRECATION: ":" is not allowed in test name "Coulomb: point (3D)", it has been replaced with "_"
> TESTSUITE/meson.build:123: DEPRECATION: ":" is not allowed in test name "Coulomb: GFN1 (0D)", it has been replaced with "_"
> TESTSUITE/meson.build:124: DEPRECATION: ":" is not allowed in test name "Coulomb: GFN1 (3D)", it has been replaced with "_"
> TESTSUITE/meson.build:125: DEPRECATION: ":" is not allowed in test name "Coulomb: GFN2 (0D)", it has been replaced with "_"
> TESTSUITE/meson.build:126: DEPRECATION: ":" is not allowed in test name "Coulomb: GFN2 (3D)", it has been replaced with "_"
> TESTSUITE/meson.build:127: DEPRECATION: ":" is not allowed in test name "Coulomb: Gauss. (0D)", it has been replaced with "_"
> TESTSUITE/meson.build:128: DEPRECATION: ":" is not allowed in test name "Coulomb: Gauss. (3D)", it has been replaced with "_"
> TESTSUITE/meson.build:130: DEPRECATION: ":" is not allowed in test name "EEQ model: water", it has been replaced with "_"
> TESTSUITE/meson.build:131: DEPRECATION: ":" is not allowed in test name "EEQ model: 3D Ewald", it has been replaced with "_"
> TESTSUITE/meson.build:132: DEPRECATION: ":" is not allowed in test name "EEQ model: GBSA", it has been replaced with "_"
> TESTSUITE/meson.build:133: DEPRECATION: ":" is not allowed in test name "EEQ model: GBSA (salt)", it has been replaced with "_"
> TESTSUITE/meson.build:134: DEPRECATION: ":" is not allowed in test name "EEQ model: GBSA (H-bond)", it has been replaced with "_"
> TESTSUITE/meson.build:136: DEPRECATION: ":" is not allowed in test name "CN: latticepoints", it has been replaced with "_"
> TESTSUITE/meson.build:137: DEPRECATION: ":" is not allowed in test name "CN: neighbourlist", it has been replaced with "_"
> TESTSUITE/meson.build:138: DEPRECATION: ":" is not allowed in test name "DFT-D3: latticepoints", it has been replaced with "_"
> TESTSUITE/meson.build:139: DEPRECATION: ":" is not allowed in test name "DFT-D3: neighbourlist", it has been replaced with "_"
> TESTSUITE/meson.build:140: DEPRECATION: ":" is not allowed in test name "DFT-D3: threebody (LP)", it has been replaced with "_"
> TESTSUITE/meson.build:141: DEPRECATION: ":" is not allowed in test name "DFT-D3: threebody (NL)", it has been replaced with "_"
> TESTSUITE/meson.build:142: DEPRECATION: ":" is not allowed in test name "DFT-D4: latticepoints", it has been replaced with "_"
> TESTSUITE/meson.build:143: DEPRECATION: ":" is not allowed in test name "DFT-D4: neighbourlist", it has been replaced with "_"
> TESTSUITE/meson.build:144: DEPRECATION: ":" is not allowed in test name "DFT-D4: threebody (LP)", it has been replaced with "_"
> TESTSUITE/meson.build:145: DEPRECATION: ":" is not allowed in test name "DFT-D4: threebody (NL)", it has been replaced with "_"
> TESTSUITE/meson.build:147: DEPRECATION: ":" is not allowed in test name "GFN2-xTB: Basic", it has been replaced with "_"
> TESTSUITE/meson.build:148: DEPRECATION: ":" is not allowed in test name "GFN2-xTB: Solvation", it has been replaced with "_"
> TESTSUITE/meson.build:149: DEPRECATION: ":" is not allowed in test name "GFN2-xTB: SCC", it has been replaced with "_"
> TESTSUITE/meson.build:150: DEPRECATION: ":" is not allowed in test name "GFN2-xTB: API", it has been replaced with "_"
> TESTSUITE/meson.build:151: DEPRECATION: ":" is not allowed in test name "GFN2-xTB: API (GBSA)", it has been replaced with "_"
> TESTSUITE/meson.build:152: DEPRECATION: ":" is not allowed in test name "GFN2-xTB: API (GBSA+salt)", it has been replaced with "_"
> TESTSUITE/meson.build:153: DEPRECATION: ":" is not allowed in test name "GFN2-xTB: d-Metal", it has been replaced with "_"
> TESTSUITE/meson.build:167: DEPRECATION: ":" is not allowed in test name "GFN1-xTB: Basic", it has been replaced with "_"
> TESTSUITE/meson.build:168: DEPRECATION: ":" is not allowed in test name "GFN1-xTB: Solvation", it has been replaced with "_"
> TESTSUITE/meson.build:169: DEPRECATION: ":" is not allowed in test name "GFN1-xTB: SCC", it has been replaced with "_"
> TESTSUITE/meson.build:170: DEPRECATION: ":" is not allowed in test name "GFN1-xTB: API", it has been replaced with "_"
> TESTSUITE/meson.build:171: DEPRECATION: ":" is not allowed in test name "GFN1-xTB: XB", it has been replaced with "_"
> TESTSUITE/meson.build:172: DEPRECATION: ":" is not allowed in test name "GFN1-xTB: PBC", it has been replaced with "_"
> TESTSUITE/meson.build:173: DEPRECATION: ":" is not allowed in test name "GFN1-xTB: API (GBSA)", it has been replaced with "_"
> TESTSUITE/meson.build:174: DEPRECATION: ":" is not allowed in test name "GFN1-xTB: API (PCEM)", it has been replaced with "_"
> TESTSUITE/meson.build:175: DEPRECATION: ":" is not allowed in test name "GFN1-xTB: IPEA", it has been replaced with "_"
> TESTSUITE/meson.build:177: DEPRECATION: ":" is not allowed in test name "GFN0-xTB: Basic", it has been replaced with "_"
> TESTSUITE/meson.build:178: DEPRECATION: ":" is not allowed in test name "GFN0-xTB: Solvation", it has been replaced with "_"
> TESTSUITE/meson.build:179: DEPRECATION: ":" is not allowed in test name "GFN0-xTB: SP", it has been replaced with "_"
> TESTSUITE/meson.build:180: DEPRECATION: ":" is not allowed in test name "GFN0-xTB: API", it has been replaced with "_"
> TESTSUITE/meson.build:181: DEPRECATION: ":" is not allowed in test name "GFN0-xTB: SRB", it has been replaced with "_"
> TESTSUITE/meson.build:182: DEPRECATION: ":" is not allowed in test name "GFN0-xTB: SP  (PBC)", it has been replaced with "_"
> TESTSUITE/meson.build:183: DEPRECATION: ":" is not allowed in test name "GFN0-xTB: API (PBC)", it has been replaced with "_"
> TESTSUITE/meson.build:184: DEPRECATION: ":" is not allowed in test name "GFN0-xTB: SRB (PBC)", it has been replaced with "_"
> TESTSUITE/meson.build:186: DEPRECATION: ":" is not allowed in test name "GFN-FF: Basic", it has been replaced with "_"
> TESTSUITE/meson.build:187: DEPRECATION: ":" is not allowed in test name "GFN-FF: Solvation", it has been replaced with "_"
> TESTSUITE/meson.build:188: DEPRECATION: ":" is not allowed in test name "GFN-FF: SP", it has been replaced with "_"
> TESTSUITE/meson.build:189: DEPRECATION: ":" is not allowed in test name "GFN-FF: HB", it has been replaced with "_"
> TESTSUITE/meson.build:190: DEPRECATION: ":" is not allowed in test name "GFN-FF: GBSA", it has been replaced with "_"
> TESTSUITE/meson.build:191: DEPRECATION: ":" is not allowed in test name "GFN-FF: Scale-up", it has been replaced with "_"
> TESTSUITE/meson.build:192: DEPRECATION: ":" is not allowed in test name "GFN-FF: PDB", it has been replaced with "_"
> TESTSUITE/meson.build:193: DEPRECATION: ":" is not allowed in test name "GFN-FF: SDF", it has been replaced with "_"
> Build targets in project: 5
> 
> Found ninja-1.10.2 at /home/chris/miniconda3/bin/ninja
>                                                                                                     
> ERROR: Linker pgi does not support link_whole
> 
> A full log can be found at /home/chris/Downloads/xtb/xtb/build_gpu/meson-logs/meson-log.txt
> ninja: Entering directory `build'
> ninja: error: loading 'build.ninja': No such file or directory

</details>

Any ideas on how to solve this?
Thanks.

**Reply by awvwgk** [reply]
Try https://github.com/grimme-lab/xtb/pull/407

**Reply by Chemistza74** [reply]
The compilation gets further, but I still get errors:

> Build targets in project: 3
> 
> Found ninja-1.10.2 at /home/chris/miniconda3/bin/ninja
> ninja: Entering directory `build'                                                                   
> [1/541] Compiling Fortran object lib_xtb.a.p/src_mctc_mctc_strings.f90.o
> FAILED: lib_xtb.a.p/src_mctc_mctc_strings.f90.o 
> nvfortran -Ilib_xtb.a.p -I. -I.. -I../include -Minform=inform -O2 -g -fPIC -module lib_xtb.a.p -o lib_xtb.a.p/src_mctc_mctc_strings.f90.o -c ../src/mctc/mctc_strings.f90
> NVFORTRAN-I-0035-Predefined intrinsic loc loses intrinsic property (../src/mctc/mctc_strings.f90: 343)
> NVFORTRAN-S-0026-Unmatched quote (../src/mctc/mctc_strings.f90: 835)
>   0 inform,   0 warnings,   1 severes, 0 fatal for split
> NVFORTRAN-S-0026-Unmatched quote (../src/mctc/mctc_strings.f90: 896)
>   0 inform,   0 warnings,   1 severes, 0 fatal for removebksl
> [8/541] Compiling Fortran object lib_xtb.a.p/src_mctc_version.f90.o
> NVFORTRAN-I-0035-Predefined intrinsic date loses intrinsic property (./xtb_version.fh: 2)
> [11/541] Compiling Fortran object lib_xtb.a.p/src_mctc_chartools.f90.o
> NVFORTRAN-I-0035-Predefined intrinsic ilen loses intrinsic property (../src/mctc/chartools.f90: 44)
> [17/541] Compiling Fortran object lib_xtb.a.p/src_type_reader.F90.o
> NVFORTRAN-I-0035-Predefined intrinsic size loses intrinsic property (../src/type/reader.F90: 131)
> [20/541] Compiling Fortran object lib_xtb.a.p/src_mctc_systools.F90.o
> NVFORTRAN-I-0035-Predefined intrinsic size loses intrinsic property (../src/mctc/systools.F90: 45)
> [26/541] Compiling Fortran object lib_xtb.a.p/src_cqpath.f90.o
> ninja: build stopped: subcommand failed.
> 
Many thanks.

**Reply by awvwgk** [reply]
Looks like the `-Mbackslash` option is missing, try adding `-Dfortran_args=-Mbackslash` when setting up the build with meson.

**Reply by Chemistza74** [reply]
That did the trick! XTB now compiles and runs. However, the output I'm getting for a geometry optimisation is garbage:

> ........................................................................
> .............................. CYCLE    1 ..............................
> ........................................................................
>    1    -62.4069560 -0.624070E+02  0.126E-06    9.76       0.0  T
>    2    -62.4069560  0.284217E-13  0.751E-07    9.76   20558.4  T
>    3    -62.4069560  0.710543E-14  0.168E-07    9.76   91839.1  T
>      SCC iter.                  ...        0 min,  0.012 sec
>      gradient                   ...        0 min,  0.022 sec
>  * total energy  :   -61.6755654 Eh     change       -0.2842171E-13 Eh
>    gradient norm :           NaN Eh/α   predicted     0.0000000E+00 (-100.00%)
>    displ. norm   :           NaN α      lambda        0.0000000E+00
>    maximum displ.:     0.0000000 α      in ANC's #0, #0, #0, ...
> 
> ........................................................................
> .............................. CYCLE    2 ..............................
> ........................................................................
>    1            NaN           NaN        NaN     NaN       0.0  T
>    2            NaN           NaN        NaN     NaN  100000.0  T
>    3            NaN           NaN        NaN     NaN  100000.0  T
>    4            NaN           NaN        NaN     NaN  100000.0  T
>    5            NaN           NaN        NaN     NaN  100000.0  T
>    6            NaN           NaN        NaN     NaN  100000.0  T
>    7            NaN           NaN        NaN     NaN  100000.0  T

It looks like it's calculating the correct energy, but gradient=NaN. Are there any other settings or options that I'm not including?

My build config looks like this:

> export FC=nvfortran CC=nvc
> meson setup build -Dla_backend=openblas -Dfortran_args=-Mbackslash -Dopenmp=false -Dgpu=true -Dcusolver=true -Dgpu_arch=86

I have a new RTX 3060TI GPU which I believe is gpu_arch=86 (I've also tried gpu_arch=70, with the same result).

Thanks again!

**Reply by awvwgk** [reply]
So far I used the GPU version only on Tesla V100 and Tesla K80 cards, so no experience with RTX cards up to now.

**Comment by awvwgk** [comment]
I added a short introduction to compiling `xtb` with NVHPC here:
https://xtb-docs.readthedocs.io/en/latest/development.html#building-with-gpu-support

**Comment by vsinha027** [comment]
Hi,
I am trying to compile xtb to be used on GPUs as well.
I followed the steps in the documentation.
The installation seemed to go fine.
But when I run the code I get the following error message:

xtb: symbol lookup error: /sw/arch/RedHatEnterpriseServer7/EB_production/2020/software/NVHPC/20.7/Linux_x86_64/20.7/compilers/lib/liblapack.so.0: undefined symbol: pgf90_str_copy_klen


Not sure how to handle this. Help/suggestions appreciated.

**Reply by awvwgk** [reply]
Looks like you might have not loaded the runtime libraries anymore in your environment.

**Reply by vsinha027** [reply]
the NVHPC module is loaded, I can also see it in the LD_LIBRARY_PATH
The xtb module is also loaded in the runtime.

Btw when I compiled I did it on the login node which is a CPU node. Could that be an issue?

Or are there more libraries to be loaded.

**Reply by vsinha027** [reply]
The GPUs are Tesla K40m, may be that is causing an issue.
